src/parquet/arrow/reader.cc

```diff
@@ -868,12 +870,227 @@ struct TransferFunctor<
+    const int64_t length = fixed_size_binary_array.length();
+
+    // allocate memory for the decimal array
+    std::shared_ptr<Buffer> data;
+    RETURN_NOT_OK(::arrow::AllocateBuffer(pool, length * type_length, &data));
+
+    // raw bytes that we can write to
+    uint8_t* out_ptr = data->mutable_data();
+
+    // convert each FixedSizeBinary value to valid decimal bytes
+    const int64_t null_count = fixed_size_binary_array.null_count();
+    if (null_count > 0) {
+      for (int64_t i = 0; i < length; ++i, out_ptr += type_length) {
+        if (!fixed_size_binary_array.IsNull(i)) {
+          RawBytesToDecimalBytes(fixed_size_binary_array.GetValue(i), byte_width,
+                                 out_ptr);
+        }
+      }
+    } else {
+      for (int64_t i = 0; i < length; ++i, out_ptr += type_length) {
+        RawBytesToDecimalBytes(fixed_size_binary_array.GetValue(i), byte_width, out_ptr);
+      }
+    }
+
+    *out = std::make_shared<::arrow::Decimal128Array>(
+        type, length, data, fixed_size_binary_array.null_bitmap(), null_count);
+    return Status::OK();
+  }
+};
+
+/// \brief Convert an Int32 or Int64 array into a Decimal128Array
+/// The parquet spec allows systems to write decimals in int32, int64 if the values are
+/// small enough to fit in less 4 bytes or less than 8 bytes, respectively.
+/// This function implements the conversion from int32 and int64 arrays to decimal arrays.
+template <typename ParquetIntegerType,
+          typename = typename std::enable_if<
+              std::is_same<ParquetIntegerType, Int32Type>::value ||
+              std::is_same<ParquetIntegerType, Int64Type>::value>::type>
+static Status DecimalIntegerTransfer(RecordReader* reader, MemoryPool* pool,
+                                     const std::shared_ptr<::arrow::DataType>& type,
+                                     std::shared_ptr<Array>* out) {
+  DCHECK_EQ(type->id(), ::arrow::Type::DECIMAL);
+
+  const int64_t length = reader->values_written();
+
+  using ElementType = typename ParquetIntegerType::c_type;
+  static_assert(std::is_signed<ElementType>::value, "ElementType is not signed");
+
+  const auto values = reinterpret_cast<const ElementType*>(reader->values());
+
+  const auto& decimal_type = static_cast<const ::arrow::DecimalType&>(*type);
+  const int64_t type_length = decimal_type.byte_width();
+
+  std::shared_ptr<Buffer> data;
+  RETURN_NOT_OK(::arrow::AllocateBuffer(pool, length * type_length, &data));
+  uint8_t* out_ptr = data->mutable_data();
+
+  using ::arrow::BitUtil::ToLittleEndian;
+
+  for (int64_t i = 0; i < length; ++i, out_ptr += type_length) {
+    const ElementType value = values[i];
+
+    auto out_ptr_view = reinterpret_cast<uint64_t*>(out_ptr);
+    out_ptr_view[0] = ToLittleEndian(static_cast<uint64_t>(value));
+
+    // no need to byteswap here because we're either all ones or all zeros
+    out_ptr_view[1] = static_cast<uint64_t>(value < 0 ? -1 : 0);
```

Comment:
The sign bit is in the same place in both cases (big vs little endian)? I'm clearly out of my depth on these details

Reply:
Thanks for pointing this out. There are actually two bugs here, revisiting this after reading your comment and rereading my code.

1. I should be calling `FromLittleEndian` just before this line, not `ToLittleEndian` because parquet-mr writes all primitive values in little endian order. The current code works on little endian architectures but not on big endian.
2. `value` needs to be sign/zero extended if it's of type `int32_t`. That's done by simply upcasting to `int64_t`.

This particular line is performing sign/zero extension to the other 8 bytes of the 16 byte decimal value, which means it's either 64 ones if `value` is negative, or 64 zeros if it's zero or positive.

Let me know if this explanation doesn't make sense.

Reply:
makes sense at a glance

Reply:
I will add an example in the code to make this very concrete. It's not necessarily obvious unless one is familiar with the details of sign extension (which is a consequence of using a two's complement representation).